src/lib/FreeC/Pass/KindInferencePass.hs

```diff
@@ -0,0 +1,211 @@
+
+-- | This module contains a compiler pass that checks if all type constructors
+--   in type expressions are fully applied and type variables are not applied.
+--
+--   For simplification, all type variables have kind @*@. Therefore, all type
+--   constructors must be fully applied, so partial application of type
+--   constructors is not allowed. This pass also checks that type variables and
+--   function types do not occur on the left hand side of a type application.
+--   If a type expression is found that does not match this rules, a fatal
+--   error is thrown. For simplification, a type expression that satisfies all
+--   of these rules is called kind-correct.
+--
+--   = Example
+--
+--   The following module contains three invalid definitions
```

Comment:
```suggestion
--   The following module contains three invalid definitions.
```

You could maybe also add a short explanation for each declaration why it is invalid.